### PR TITLE
Provide error message for malformed annotations.

### LINKED ISSUE #59113: kubectl-apply silently drops annotations with boolean values
/kind bug

**What happened**:
When creating resource with boolean value in annotation, kubectl apply completes sucessfully without any errors, but resource has no annotations.

**What you expected to happen**:

Error similar to what `kubectl create` returns

**How to reproduce it (as minimally and precisely as possible)**:

```
$ kubectl apply -f <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  annotations:
    bool: true
data: {}
EOF
configmap "test" created
```

Creates:
```
apiVersion: v1
kind: ConfigMap
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{},"kind":"ConfigMap","metadata":{"annotations":{},"name":"test","namespace":"kube-system"}}
  creationTimestamp: 2018-01-31T14:23:42Z
  name: test
  namespace: kube-system
  resourceVersion: "9612155"
  selfLink: /api/v1/namespaces/kube-system/configmaps/test
  uid: 5689a651-0692-11e8-a6e3-005056bced2e
```

**Environment**:
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T10:09:24Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:17:43Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

### PR DESCRIPTION
**Pull request by code-sleuth**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
This PR adds provision to error out rather than to silently drop the malformed annotations.

**Which issue(s) this PR fixes**:
Fixes #59113

**Special notes for your reviewer**:

/sig api-machinery
/assign @liggitt 

**Does this PR introduce a user-facing change?**:
Create a test ConfigMap file `bool-annot.yaml` with bool annotations
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  annotations:
    bool: true
data: {}
```
Run: `kubectl apply -f ./bool-annot.yaml`

Error output:
```release-note
error: error when retrieving modified configuration from:
Resource: "/v1, Resource=configmaps", GroupVersionKind: "/v1, Kind=ConfigMap"
Name: "test", Namespace: "default"
Object: &{map["apiVersion":"v1" "data":map[] "kind":"ConfigMap" "metadata":map["annotations":map["bool":%!q(bool=true)] "name":"test" "namespace":"default"]]}
for: "/root/bool-annot.yaml": .metadata.annotations accessor error: contains non-string key in the map: true is of the type bool, expected string
```

### CODE REVIEW

File: pkg/fieldpath/fieldpath.go [hunk 1]
```diff
@@ -55,7 +55,8 @@ func ExtractFieldPathAsString(obj interface{}, fieldPath string) (string, error)
 			if errs := validation.IsQualifiedName(strings.ToLower(subscript)); len(errs) != 0 {
 				return "", fmt.Errorf("invalid key subscript in %s: %s", fieldPath, strings.Join(errs, ";"))
 			}
-			return accessor.GetAnnotations()[subscript], nil
+			annotations, _ := accessor.GetAnnotations()
```
Comment (pontiyaraja) [inline pkg/fieldpath/fieldpath.go]: `accessor.GetAnnotations()` is returning only map

**Review by pontiyaraja — COMMENTED** [review]
resolve the issues, the same issue is there in multiple files. if this PR is in work in progress add `WIP` .

**Comment by mbohlool** [comment]
@logicalhan

**Comment by noqcks** [comment]
hi @code-sleuth, anything I can do to help with this PR? 

I just ran into an issue of silently failing annotations myself.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go [hunk 2]
```diff
@@ -53,7 +53,7 @@ type Object interface {
 	SetDeletionGracePeriodSeconds(*int64)
 	GetLabels() map[string]string
 	SetLabels(labels map[string]string)
-	GetAnnotations() map[string]string
+	GetAnnotations() (map[string]string, error)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go]: @smarterclayton thoughts on making these capable of returning errors? I'd want to do it consistently for all the methods, not piece-mill like this, but that's a huge change

**Comment by code-sleuth** [comment]
hi @noqcks, not really. I just have to make the same changes to `kustomize` [here](https://github.com/kubernetes/kubernetes/blob/cc67ccfd7f4f0bc96d7f1c8e5fe8577821757d03/vendor/sigs.k8s.io/kustomize/pkg/ifc/ifc.go#L60) and ripple through the other affected files too.
But lets wait and see what @liggitt and @smarterclayton agree on as the best way forward.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go (hunk 2 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go]: This isn't the right place to add this.  What you really want is a variant of Accessor https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/meta/meta.go#L102 that returns an error if you're not a metadata object.  Determining that you're a metadata object is a duck check for unstructured content, but you should be able to do that before you get here.

Comment (smarterclayton) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go]: An object that is a metadata.Object is required by law to support annotations, so it would be inappropriate to have an optional error on that interface.  You just shouldn't be able to get an Object for whatever type you need.

File: pkg/kubectl/cmd/apply/apply.go [hunk 3]
```diff
@@ -387,6 +387,7 @@ func (o *ApplyOptions) Run() error {
 				if o.ServerDryRun {
 					options.DryRun = []string{metav1.DryRunAll}
 				}
+
```
Comment (oomichi) [inline pkg/kubectl/cmd/apply/apply.go]: nit: unrelated change

File: pkg/fieldpath/fieldpath.go [hunk 4]
```diff
@@ -68,7 +69,8 @@ func ExtractFieldPathAsString(obj interface{}, fieldPath string) (string, error)
 
 	switch fieldPath {
 	case "metadata.annotations":
-		return FormatMap(accessor.GetAnnotations()), nil
+		annotations, _ := accessor.GetAnnotations()
```
Comment (oomichi) [inline pkg/fieldpath/fieldpath.go]: Don't we need to add error handling on most caller sides of GetAnnotations()?

File: staging/src/k8s.io/apimachinery/pkg/api/meta/meta.go [hunk 5]
```diff
@@ -335,7 +336,11 @@ func (resourceAccessor) Annotations(obj runtime.Object) (map[string]string, erro
 	if err != nil {
 		return nil, err
 	}
-	return accessor.GetAnnotations(), nil
+	accessorAnnotations, err := accessor.GetAnnotations()
+	if err != nil {
+		return nil, err
+	}
+	return accessorAnnotations, nil
```
Comment (oomichi) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/meta.go]: I think we can change line 339-343 into a single line like:
```
return return accessor.GetAnnotations()
```
